### ISSUE TITLE
cloud-init upstream setup.py packaging is incompatible with PEP517 distro implementations

### ISSUE THREAD

**Issue by holmanb** [canonical/cloud-init#5027]
# Bug report

tl;dr: Changing cloud-init's packaging paradigm to support PEP517 packaging tooling appears unavoidable to support PEP517.

In https://github.com/canonical/cloud-init/commit/a476ecb5b1dd47527c907d2736cb2dd9ef845119 we tried switching debian packaging from dh-python's pybuild distutils plugin, (which depends on a standard library package which was removed from stdlib in 3.12) to dh-python's pybuild pyproject plugin.

After this change, tests on this branch started failing due to [ds-identify and other "data_files"](https://github.com/canonical/cloud-init/blob/621c522ccf2897ff4d19a8ed332c187516eccf9b/setup.py#L341) being installed in the wrong location.

After [filing a bug with debian against dh-python](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1065441), I [had an IRC conversation](https://gist.github.com/holmanb/792e966782830c75711147df04401ec9) in #debian-python on Libera. To summarize the conversation: other non-debian distros' packaging tooling similarly does wheel-based installations, so it's unlikely that we'll be able to keep the current packaging paradigm.

**Comment by dermotbradley** [comment]
I had discussed this in the #cloud-init IRC channel several months ago as Alpine had switched a lot of packages from setup.py to gpep517.

> After this change, tests on this branch started failing due to [ds-identify and other "data_files"](https://github.com/canonical/cloud-init/blob/621c522ccf2897ff4d19a8ed332c187516eccf9b/setup.py#L341) being installed in the wrong location.

That was exactly the same issue I experienced then when I tried to switch the Alpine cloud-init package to gpep517. At the time (from memory) Sam (_sam on IRC) indicated that gpep517 didn't support files outside of /var/lib/python/site-packages/ currently

**Comment by holmanb** [comment]
> I had discussed this in the #cloud-init IRC channel several months ago as Alpine had switched a lot of packages from setup.py to gpep517.
> 
> > After this change, tests on this branch started failing due to [ds-identify and other "data_files"](https://github.com/canonical/cloud-init/blob/621c522ccf2897ff4d19a8ed332c187516eccf9b/setup.py#L341) being installed in the wrong location.
> 
> That was exactly the same issue I experienced then when I tried to switch the Alpine cloud-init package to gpep517. At the time (from memory) Sam (_sam on IRC) indicated that gpep517 didn't support files outside of /var/lib/python/site-packages/ currently
> 

@dermotbradley +1 Thanks for adding the extra context. My understanding is that distro packaging will be required to relocate from the site-packages prefix after using 517 tooling.

**Comment by eli-schwartz** [comment]
The python ecosystem officially considers it unsupported to use python packaging technology for anything other than:
- *.py importable libraries of code
- command-line entry scripts

The wheel standard supports installing files to hardcoded locations inside of /usr if and only if sys.prefix for python itself is /usr -- this means that installing into a virtualenv is probably broken as I somehow doubt the software will work correctly when installing `/home/myname/.venvmanager/venvs/cloud-init/lib/udev/*`, at which point you really, really should be asking "what am I getting out of this".

Note also that it is IMPOSSIBLE to install files to /etc using a wheel, since /etc inherently lies outside of python's sys.prefix.

On the flip side, you need to ask what you actually get out of wheels. It allows users to pip install the software, but do you need that functionality? Probably not... just get the software from the distro repository.

Effectively, python packaging is being used as a glorified makefile, and setup.py is kind of convenient because it supports locating the system site-packages directory.

I would recommend ditching the python packaging ecosystem entirely if you don't need it, and moving to a real build system such as [meson](https://mesonbuild.com/).

You still get support for installing python modules using `py_installation.install_sources()`, but you also get to install files to standard locations such as --prefix, --sysconfdir, --localstatedir, and so on.

**Comment by holmanb** [comment]
This has been added to the 24.3 project milestone. FYI @dermotbradley @prometheanfire @thesamesam @rjschwei @igalic @ani-sinha @ncopa @major @nmeyerhans

**Comment by prometheanfire** [comment]
Gentoo is now going to remove cloud-init as a package. https://bugs.gentoo.org/909872

**Comment by thesamesam** [comment]
I've given it a reprieve in Gentoo by hacking around it for now, but there's still a hard deadline of whenever setuptools finally drops the `setup.py` frontend/entrypoint for installing.

**Comment by thesamesam** [comment]
In setuptools-80.1.0, they've finally [set a removal date](https://github.com/pypa/setuptools/commit/7fc5e05df9e239cf2938396dcb1ca93d00621e2a) for 2025-10-31 for `setup.py install`.

**Comment by TheRealFalcon** [comment]
Just a heads up that this is actively being worked on. It just got delayed due to some shifting priorities.

**Comment by dermotbradley** [comment]
@TheRealFalcon FYI from setuptools v80.1.0 [NEWS.rts](https://github.com/pypa/setuptools/blob/main/NEWS.rst#features-6):

> Added a deadline of Oct 31 to the setup.py install deprecation.

**Comment by thesamesam** [comment]
That's what I said in my comment above.

**Comment by TheRealFalcon** [comment]
A WIP PR for this can be found at https://github.com/canonical/cloud-init/pull/6236
